src/transformer_util.ts

```diff
@@ -0,0 +1,545 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as ts from 'typescript';
+import * as tsickle from './tsickle';
+
+/**
+ * Adjusts the given CustomTransformers with additional transformers
+ * to fix bugs in TypeScript.
+ * @param given A
+ */
+export function createCustomTransformers(given: ts.CustomTransformers): ts.CustomTransformers {
+  if (!given.after && !given.before) {
+    return given;
+  }
+  const before = given.before || [];
+  before.unshift(firstTransform);
+  before.push(preTypeScriptTransform);
+  const after = given.after || [];
+  after.unshift(postTypescriptTransform);
+  return {before, after};
+}
+
+/**
+ * Transform to be used as first transform to reset some state.
+ */
+function firstTransform(context: ts.TransformationContext) {
+  return (sourceFile: ts.SourceFile) => {
+    (context as TransformationContext).fileContext = new FileContext(sourceFile);
```

Comment:
Document that all transformers run in sequence for each file before visiting the next file, which makes this operation (caching the source file here) valid.